connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -117,6 +119,10 @@ private[sql] class AvroDeserializer(
     val incompatibleMsg = errorPrefix +
         s"schema is incompatible (avroType = $avroType, sqlType = ${catalystType.sql})"
 
+    val realDataType = SchemaConverters.toSqlType(avroType).dataType
+    val confKey = SQLConf.LEGACY_AVRO_ALLOW_INCOMPATIBLE_SCHEMA
+    val preventReadingIncorrectType = !SQLConf.get.getConf(confKey)
+
     (avroType.getType, catalystType) match {
```

Comment:
After thinking about it more, I think the current code structure is fragile and not future-proof. What if we have a new logical type based on avro BOOLEAN and we don't want to read it as a boolean type? The current code structure matches the avro physical type and the read catalyst type, and we only check the corresponding catalyst type of the avro type for certain cases.

Instead of using a deny list, I think an allow list is more future-proof: We explicitly list all the cases that we allow a mismatch of the catalyst type. For example, we can read a long type avro value as timestamp type for backward compatibility reasons.

Reply:
cc @gengliangwang

Reply:
Changed the structure now!

Reply:
Hi @gengliangwang and @cloud-fan, I receive a ticket about the performance slowdown on avro reader because of this PR. https://databricks.atlassian.net/browse/ES-761029. In this PR, to make code more readable, we turn one match-case statement into nested match-case statement which slows down the AvroDeserializer. Do you think we should revert this code structure?

Reply:
@zeruibao do you know the exact reason of the slowdown? The schema conversion is one-time thing on the driver.

Reply:
If the avro file is really big. It will still cause performance regression.

Reply:
The test file is a big avro containing 1000000 rows and 300 columns

Reply:
@zeruibao have you confirmed the perf is restored after reverting the code changes in this function? If yes, let's restore to the previous pattern match.

Reply:
Yes, the perf is restored after reverting this commit. And for this commit, I only change this function (except test suite).  So, reverting to previous pattern match should be good.